singularity-eos/base/root-finding-1d/root_finding.hpp

```diff
@@ -235,6 +237,71 @@ PORTABLE_INLINE_FUNCTION Status regula_falsi(const T &f, const Real ytarget,
   return status;
 }
 
+// solves for f(x,params) - ytarget = 0
+template <typename T>
+PORTABLE_INLINE_FUNCTION Status newton_raphson(const T &f, const Real ytarget,
+                                               const Real guess, const Real a,
+                                               const Real b, const Real ytol, Real &xroot,
+                                               const RootCounts *counts = nullptr,
+                                               const bool &verbose = false) {
+
+  constexpr int max_iter = NEWTON_RAPHSON_NITER_MAX;
+  Real _x = guess;
+  Real _xold = 0.0;
+  auto status = Status::SUCCESS;
+
+  Real yg;
+  Real dfunc;
+
+  int iter;
+
+  for (iter = 0; iter < max_iter; iter++) {
+    std::tie(yg, dfunc) = f(_x); // C++11 tuple unpacking
+
+    // check if we are converged already
+    if (std::abs(yg - ytarget) < (ytol * ytarget)) break;
+
+    // not converged; compute the next step
+    _xold = _x;
+    _x = _x - (yg - ytarget) / dfunc;
+
+    // check if we are out of bounds
+    if (_x <= a && _xold <= a) {
+      if (verbose) {
+        printf("newton_raphson out of bounds! %.14e %.14e %.14e %.14e\n", ytarget, guess,
+               a, b);
+      }
+      break;
+    }
+    if (_x >= b && _xold >= b) {
+      if (verbose) {
+        printf("newton_raphson out of bounds! %.14e %.14e %.14e %.14e\n", ytarget, guess,
+               a, b);
+      }
+      break;
+    }
```

Comment:
Maybe I'm missing the difference, but it looks like these conditions could be combined to make this code more DRY

Reply:
Thanks for the catch! This is a bit of a tricky one. In the original reference implementation, in each condition the root value was set to the appropriate boundary value, hence the two conditions. This has been moved to the Helmholtz call itself (since it is also necessary for the RF roots). I'm not sure if this is the smart way to do, i.e. if this bounding should be handled by the root finding methods or by the function calling the root find. For now I have decided that this should be handled by the function calling the root find (since one might not always want to bound the value to the boundary values) and added a comment in the appropriate location.

This is also the reason that this out-of-bounds case does not return a FAILURE status. One could argue that this is not really a successful root, but at least in the context of the Helmholtz EOS a root bound to the edges of the table is (at least in the reference implementation) taken as a success. Most likely this is a peculiarity for the Helmholtz EOS due to the fine tuning and won't translate to other application. I am however hesitant to implement a FAILURE return here since this would trigger an additional root find using RF. The bound result should be good enough, at least good enough as to not trigger a new attempt.

Reply:
Fair point. Thanks for the suggestion! I've added the relevant option to the root find.